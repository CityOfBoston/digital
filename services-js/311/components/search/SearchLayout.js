// @flow

import React from 'react';
import type { Context } from 'next';
import Router from 'next/router';
import { css } from 'glamor';
import { action, reaction } from 'mobx';
import { observer } from 'mobx-react';
import Head from 'next/head';

import type { RequestAdditions } from '../../server/next-handlers';
import type { AppStore } from '../../data/store';

import makeLoopbackGraphql from '../../data/dao/loopback-graphql';
import type { LoopbackGraphql } from '../../data/dao/loopback-graphql';

import Nav from '../common/Nav';
import LocationMap from '../map/LocationMap';
import { MEDIA_LARGE, HEADER_HEIGHT } from '../style-constants';

import RecentRequests from './RecentRequests';

type SearchData = {|
  view: 'search',
  query: string,
  location: ?{lat: number, lng: number},
  zoom: ?number,
|}

export type InitialProps = {|
  data: SearchData,
|};

export type Props = {|
  /* :: ...InitialProps, */
  store: AppStore,
|}

// This is the main container for content. We want to be at least full height on
// large screens to push the footer down to where you need to scroll for it.
const CONTAINER_STYLE = css({
  minHeight: 0,
  position: 'relative',
  justifyContent: 'flex-start',
  [MEDIA_LARGE]: {
    minHeight: '100vh',
  },
});

const BACKGROUND_MAP_CONTAINER_STYLE = css({
  position: 'fixed',
  width: '100%',
  top: HEADER_HEIGHT,
  bottom: 0,
  background: '#9B9B9B',
});

@observer
export default class LookupLayout extends React.Component {
  props: Props;
  loopbackGraphql: LoopbackGraphql = makeLoopbackGraphql();

  locationUpdateDisposer: ?Function;

  static async getInitialProps({ query }: Context<RequestAdditions>): Promise<InitialProps> {
    const { lat, lng, zoom } = query;
    let location;

    if (lat && lng) {
      location = {
        lat: parseFloat(query.lat),
        lng: parseFloat(query.lng),

      };
    } else {
      location = null;
    }

    return {
      data: {
        view: 'search',
        query: query.q || '',
        location,
        zoom: zoom ? parseInt(zoom, 10) : null,
      },
    };
  }

  @action
  componentWillMount() {
    const { store, data } = this.props;
    store.requestSearch.query = data.query;

    if (data.location) {
      store.requestSearch.mapCenter = data.location;
    } else {
      store.requestSearch.mapCenter = null;
    }

    if (data.zoom) {
      store.requestSearch.mapZoom = data.zoom;
    } else {
      store.requestSearch.mapZoom = 12;
    }
  }

  @action
  componentDidMount() {
    const { store } = this.props;

    store.requestSearch.start(this.loopbackGraphql);

    this.locationUpdateDisposer = reaction(
      () => ({
        resultsQuery: store.requestSearch.resultsQuery,
        location: store.requestSearch.mapCenter,
        zoom: store.requestSearch.mapZoom,
      }),
      ({ resultsQuery, location, zoom }) => {
        const params = new URLSearchParams();

        if (resultsQuery) {
          params.append('q', resultsQuery);
        }

        if (location) {
          params.append('lat', location.lat);
          params.append('lng', location.lng);
        }

        params.append('zoom', zoom);

        const href = `/search?${params.toString()}`;
        Router.push(href, href, { shallow: true });
      },
      {
        name: 'location query updater',
      },
    );
  }

  componentWillUnmount() {
    const { store } = this.props;
    store.requestSearch.stop();

    if (this.locationUpdateDisposer) {
      this.locationUpdateDisposer();
      this.locationUpdateDisposer = null;
    }
  }

  render() {
    const { store } = this.props;

    return (
      <div className={`mn mn--full mn--nv-s ${CONTAINER_STYLE.toString()}`} style={{ backgroundColor: 'transparent' }}>
        <Head>
          <title>BOS:311 — Public Cases</title>
        </Head>

        <Nav activeSection="search" />

        <div className={BACKGROUND_MAP_CONTAINER_STYLE}>
          <LocationMap
            store={store}
            mode={'requests'}
            mobile={false}
          />
        </div>

        <RecentRequests store={store} />
      </div>

    );
  }

}
