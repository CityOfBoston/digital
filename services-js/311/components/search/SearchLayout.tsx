import 'url-search-params-polyfill';

import React from 'react';
import Router from 'next/router';
import { css } from 'emotion';
import { action, reaction, observable, computed } from 'mobx';
import { observer } from 'mobx-react';
import Head from 'next/head';

import {
  MEDIA_LARGE,
  HEADER_HEIGHT,
  IPHONE_FOOTER_HEIGHT,
  WHITE,
} from '@cityofboston/react-fleet';

import RequestSearch from '../../data/store/RequestSearch';

import FeedbackBanner from '../common/FeedbackBanner';
import Footer from '../common/Footer';
import Nav from '../common/Nav';
import { LocationMapWithLibrary } from '../map/LocationMap';
import LocationMap from '../map/LocationMap';

import RecentRequests from './RecentRequests';
import RecentRequestsSearchForm from './RecentRequestsSearchForm';
import { PageDependencies, GetInitialProps } from '../../pages/_app';
import { getParam } from '@cityofboston/next-client-common';

type SearchData = {
  view: 'search';
  query: string;
  location: { lat: number; lng: number } | null;
  zoom: number | null;
};

export type InitialProps = {
  data: SearchData;
};

export type Props = InitialProps &
  Pick<
    PageDependencies,
    | 'addressSearch'
    | 'ui'
    | 'requestSearch'
    | 'browserLocation'
    | 'fetchGraphql'
  > & {
    noMap: boolean;
  };

const SEARCH_HEIGHT = 62;

// This is the main container for content. We want to be at least full height on
// large screens to push the footer down to where you need to scroll for it.
const CONTAINER_STYLE = css({
  position: 'relative',
  minHeight: `calc(100vh - ${HEADER_HEIGHT}px - ${IPHONE_FOOTER_HEIGHT}px)`,
  paddingTop: SEARCH_HEIGHT,
  display: 'flex',
  flexDirection: 'column',

  [MEDIA_LARGE]: {
    paddingTop: 0,
  },
});

const CONTAINER_FULL_MAP_STYLE = css(CONTAINER_STYLE, {
  height: `calc(100vh - ${HEADER_HEIGHT}px - ${IPHONE_FOOTER_HEIGHT}px)`,
});

const SEARCH_CONTAINER_STYLE = css({
  display: 'block',
  height: SEARCH_HEIGHT,
  width: '100%',
  position: 'fixed',
  top: HEADER_HEIGHT,
  zIndex: 2,
  background: WHITE,

  [MEDIA_LARGE]: {
    display: 'none',
  },
});

const MAP_CONTAINER_STYLE = css({
  height: '30vh',
  background: '#9B9B9B',
  position: 'relative',
  zIndex: 0,
  display: 'flex',
  flexDirection: 'column',

  [MEDIA_LARGE]: {
    position: 'fixed',
    right: 0,
    width: '60%',
    minWidth: 'calc(100% - 35rem)',
    top: HEADER_HEIGHT,
    bottom: 0,
    height: 'auto',
  },
});

const FULL_MAP_CONTAINER_STYLE = css(MAP_CONTAINER_STYLE, {
  height: 'auto',
  flex: 1,
});

const MAP_VIEW_BUTTON_CONTAINER_STYLE = css({
  background: WHITE,
});

const STICKY_VIEW_BUTTON_STYLE = css({
  position: 'fixed',
  left: 0,
  right: 0,
  bottom: 0,
});

@observer
export default class SearchLayout extends React.Component<Props> {
  static defaultProps = {
    noMap: false,
  };

  private locationUpdateDisposer: Function | null = null;
  private currentLocationMonitorDisposer: Function | null = null;

  // These are all observable because they’re used in @computed methods.
  @observable private container: HTMLElement | null = null;
  @observable private locationMap: LocationMap | null = null;
  @observable private locationMapContainer: HTMLElement | null = null;
  @observable private mapViewButtonContainer: HTMLElement | null = null;

  static getInitialProps: GetInitialProps<InitialProps, 'query'> = async (
    { query },
    _
  ): Promise<InitialProps> => {
    const { lat, lng, zoom } = query;
    let location;

    if (lat && lng) {
      location = {
        lat: parseFloat(getParam(lat)!),
        lng: parseFloat(getParam(lng)!),
      };
    } else {
      location = null;
    }

    return {
      data: {
        view: 'search',
        query: getParam(query.q, ''),
        location,
        zoom: zoom ? parseInt(getParam(zoom)!, 10) : null,
      },
    };
  };

  @action
  componentWillMount() {
    const { requestSearch, data } = this.props;
    requestSearch.query = data.query;

    if (data.location) {
      requestSearch.mapCenter = data.location;
    } else {
      requestSearch.mapCenter = null;
    }

    if (data.zoom) {
      requestSearch.mapZoom = data.zoom;
    } else {
      requestSearch.mapZoom = 12;
    }
  }

  @action
  componentDidMount() {
    const { browserLocation, requestSearch, ui, fetchGraphql } = this.props;

    requestSearch.start(fetchGraphql);

    // If the browser's location comes in while the map is still in a default
    // view, zoom in to the browser's location.
    this.currentLocationMonitorDisposer = reaction(
      () => ({
        mobile: ui.belowMediaLarge,
        browserLocation: browserLocation.location,
        inBoston: browserLocation.inBoston,
      }),
      ({ mobile, browserLocation, inBoston }) => {
        const mapCenter = requestSearch.mapCenter;

        if (!mobile || !inBoston || !browserLocation || !this.locationMap) {
          return;
        }

        if (!mapCenter || !requestSearch.mapMoved) {
          this.locationMap.visitLocation(browserLocation, true);
        }
      }
    );

    this.locationUpdateDisposer = reaction(
      () => ({
        query: requestSearch.query,
        resultsQuery: requestSearch.resultsQuery,
        location: requestSearch.mapCenter,
        zoom: requestSearch.mapZoom,
      }),
      ({ query, resultsQuery, location, zoom }) => {
        const params = new URLSearchParams();
        const caseUrls = RequestSearch.caseUrlsForSearchQuery(query);

        if (caseUrls) {
          Router.push(caseUrls.url, caseUrls.as);
        } else {
          if (resultsQuery) {
            params.append('q', resultsQuery);
          }

          if (location) {
            params.append('lat', location.lat.toString());
            params.append('lng', location.lng.toString());
          }

          params.append('zoom', zoom.toString());

          const href = `/search?${params.toString()}`;
          Router.push(href, href, { shallow: true });
        }
      },
      {
        name: 'location query updater',
        delay: 500,
      }
    );
  }

  componentWillUnmount() {
    const { requestSearch } = this.props;
    requestSearch.stop();

    if (this.locationUpdateDisposer) {
      this.locationUpdateDisposer();
      this.locationUpdateDisposer = null;
    }

    if (this.currentLocationMonitorDisposer) {
      this.currentLocationMonitorDisposer();
      this.currentLocationMonitorDisposer = null;
    }
  }

  @action.bound
  setContainer(container: HTMLElement | null) {
    this.container = container;
  }

  @action.bound
  setLocationMap(locationMap: LocationMap | null) {
    this.locationMap = locationMap;
  }

  @action.bound
  setLocationMapContainer(locationMapContainer: HTMLElement | null) {
    this.locationMapContainer = locationMapContainer;
  }

  @action.bound
  setMapViewButtonContainer(mapViewButtonContainer: HTMLElement | null) {
    this.mapViewButtonContainer = mapViewButtonContainer;
  }

  @action.bound
  switchToListView() {
    const { locationMap } = this;
    const { requestSearch } = this.props;

    requestSearch.mapView = false;

    if (locationMap) {
      requestSearch.selectedRequest = null;
      requestSearch.selectedSource = null;

      setTimeout(() => {
        locationMap.invalidateSize();
      }, 0);
    }
  }

  @action.bound
  switchToMapView() {
    const { locationMap } = this;
    const { requestSearch } = this.props;

    requestSearch.mapView = true;
    requestSearch.selectedRequest = null;

    window.scrollTo(0, 0);

    if (locationMap) {
      setTimeout(() => {
        locationMap.invalidateSize();
      }, 0);
    }
  }

  // Returns true if the visitor has scrolled down enough in the list view that
  // the map is hidden. Signal that we should show the "Switch to map view"
  // button.
  @computed
  get mapScrolledOff(): boolean {
    const { ui } = this.props;
    const { locationMapContainer } = this;

    return !!(
      ui.scrollY > -1 &&
      locationMapContainer &&
      locationMapContainer.getBoundingClientRect().bottom <
        HEADER_HEIGHT + SEARCH_HEIGHT
    );
  }

  // Returns true if the map view button should be sticky. Logic here is that
  // we've scrolled down far enough that the map is visible, but not down to
  // see the bottom of the list, where the statically-positioned button would
  // appear.
  @computed
  get stickyMapViewButton(): boolean {
    const { ui } = this.props;
    const { container, mapViewButtonContainer } = this;

    if (!container || !mapViewButtonContainer) {
      return false;
    }

    return (
      ui.scrollY > -1 &&
      container.getBoundingClientRect().bottom +
        mapViewButtonContainer.clientHeight >
        ui.visibleHeight + HEADER_HEIGHT
    );
  }

  render() {
    const {
      ui,
      requestSearch,
      browserLocation,
      addressSearch,
      noMap,
    } = this.props;
    const { mapScrolledOff, stickyMapViewButton } = this;
    const { mapView } = requestSearch;

    return (
      <div>
        <Head>
          <title>BOS:311 — Public Cases</title>
        </Head>

        <Nav activeSection="search" />

        <div
          className={
            mapView
              ? CONTAINER_FULL_MAP_STYLE.toString()
              : CONTAINER_STYLE.toString()
          }
          style={{ backgroundColor: 'transparent' }}
          role="main"
        >
          <div className={`p-a300 ${SEARCH_CONTAINER_STYLE.toString()}`}>
            <RecentRequestsSearchForm requestSearch={requestSearch} />
          </div>

          <div
            className={mapView ? FULL_MAP_CONTAINER_STYLE : MAP_CONTAINER_STYLE}
            ref={this.setLocationMapContainer}
          >
            {!noMap && (
              <LocationMapWithLibrary
                locationMapRef={this.setLocationMap}
                mode="search"
                addressSearch={addressSearch}
                browserLocation={browserLocation}
                requestSearch={requestSearch}
                ui={ui}
                mobile={ui.belowMediaLarge}
                onMapClick={ui.belowMediaLarge ? this.switchToMapView : null}
              />
            )}
          </div>

          {!mapView && (
            <div ref={this.setContainer}>
              <RecentRequests
                requestSearch={requestSearch}
                ui={ui}
                isMobile={ui.belowMediaLarge}
              />
            </div>
          )}

          {mapView && (
            <div className="g p-a300">
              <button
                type="button"
                className="btn g--12"
                onClick={this.switchToListView}
              >
                List View
              </button>
            </div>
          )}

          {!mapView && mapScrolledOff && (
            <div
              className={`g p-a300 ${MAP_VIEW_BUTTON_CONTAINER_STYLE.toString()} ${
                stickyMapViewButton ? STICKY_VIEW_BUTTON_STYLE.toString() : ''
              }`}
              ref={this.setMapViewButtonContainer}
            >
              <button
                type="button"
                className="btn g--12"
                onClick={this.switchToMapView}
              >
                Map View
              </button>
            </div>
          )}
        </div>

        <Footer />
        <FeedbackBanner fit="PAGE" />
      </div>
    );
  }
}
