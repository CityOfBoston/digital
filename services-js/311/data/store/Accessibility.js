// @flow

import { observable, autorun } from 'mobx';

export default class Accessibility {
  el: ?HTMLElement = null;

  @observable message: string = '';
  @observable interrupt: boolean = false;

  messageListenerDisposer: ?Function = null;

  attach() {
    this.el = document.getElementById('ariaLive');

    this.messageListenerDisposer = autorun(
      () => {
        if (this.el) {
          this.el.innerText = this.message;
          this.el.setAttribute(
            'aria-live',
            this.interrupt ? 'assertive' : 'polite'
          );
        }
      },
      {
        name: 'a11y message listener',
      }
    );
  }

  detatch() {
    this.el = null;

    if (this.messageListenerDisposer) {
      this.messageListenerDisposer();
    }
  }
}
