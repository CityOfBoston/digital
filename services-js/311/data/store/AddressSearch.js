// @flow

import { observable, action, computed, runInAction } from 'mobx';

import type { LoopbackGraphql } from '../dao/loopback-graphql';
import searchAddress from '../dao/search-address';
import reverseGeocode from '../dao/reverse-geocode';
import type { AddressUnit, SearchAddressPlace } from '../types';

// This class co-ordinates location queries between the LocationPopUp pane
// and the LocationMap map. LocationPopUp automatically updates the
// RequestForm (which contains the data that will go into the request) based
// on changes to this instance.
export default class AddressSearch {
  @observable query: string = '';
  @observable searching: boolean = false;
  @observable lastQuery: string = '';
  @observable.ref lastSearchError: ?Error = null;

  @observable.shallow places: ?Array<SearchAddressPlace> = null;
  @observable _currentPlaceIndex: number = -1;
  @observable highlightedPlaceIndex: number = -1;
  @observable _currentUnitIndex: number = 0;
  @observable
  currentReverseGeocodeLocation: ?{ lat: number, lng: number } = null;
  @observable currentReverseGeocodeLocationIsValid: boolean = true;

  @observable mode: 'search' | 'geocode' = 'search';

  // We keep track of whether the constituent seems to be pointing at a location
  // on the map or choosing from the location menu, since we have to choose to
  // send either lat/lng *or* address string to Open311. lat/lng matches the map
  // exactly, but only address string captures the unit.
  @observable intent: 'ADDRESS' | 'LATLNG' = 'ADDRESS';

  // Used so that LocationMap doesn't zoom things under the search box.
  searchPopupWidth: number = 0;

  loopbackGraphql: ?LoopbackGraphql;

  start(loopbackGraphql: LoopbackGraphql) {
    this.loopbackGraphql = loopbackGraphql;
  }

  stop() {
    this.loopbackGraphql = null;
  }

  @computed
  get currentPlaceIndex(): number {
    return this._currentPlaceIndex;
  }

  set currentPlaceIndex(currentPlaceIndex: number) {
    this._currentPlaceIndex = currentPlaceIndex;
    this.intent = 'ADDRESS';
  }

  @computed
  get currentUnitIndex(): number {
    return this._currentUnitIndex;
  }

  set currentUnitIndex(currentUnitIndex: number) {
    this._currentUnitIndex = currentUnitIndex;
    this.intent = 'ADDRESS';
  }

  @action
  async search(selectFirst: boolean): Promise<void> {
    if (!this.loopbackGraphql) {
      return;
    }

    this.places = null;
    this.lastQuery = this.query;
    this.lastSearchError = null;
    this.currentReverseGeocodeLocation = null;
    this.mode = 'search';

    if (!this.query) {
      return;
    }

    try {
      this.searching = true;
      const places = await searchAddress(this.loopbackGraphql, this.query);

      runInAction('search - searchAddress success', () => {
        this.searching = false;
        this.places = places;
        this.currentPlaceIndex = places.length === 1 || selectFirst ? 0 : -1;
        this.highlightedPlaceIndex = this.currentPlaceIndex;
        this.currentUnitIndex = 0;
        this.intent = 'ADDRESS';
      });
    } catch (err) {
      runInAction('search - searchAddress error', () => {
        this.searching = false;
        this.lastSearchError = err;
        window._opbeat &&
          !err.server &&
          window._opbeat('captureException', err);
      });
    }
  }

  @computed
  get location(): ?{ lat: number, lng: number } {
    return (
      this.currentReverseGeocodeLocation ||
      (this.currentPlace ? this.currentPlace.location : null)
    );
  }

  set location(location: {| lat: number, lng: number |}) {
    this.query = '';

    if (
      this.location &&
      this.location.lat === location.lat &&
      this.location.lng === location.lng
    ) {
      return;
    }

    if (!this.loopbackGraphql) {
      return;
    }

    this.currentReverseGeocodeLocation = location;
    this.places = null;
    this.mode = 'geocode';
    this.lastQuery = '';
    this.lastSearchError = null;

    this.searching = true;
    reverseGeocode(this.loopbackGraphql, location).then(
      action('reverseGeocode success', place => {
        this.searching = false;

        if (place) {
          // We overwrite the location in the place to stay with the one picked so
          // that the map marker doesnâ€™t move.
          this.places = [
            {
              location,
              address: place.address,
              addressId: place.addressId,
              units: place.units,
            },
          ];
          this.currentPlaceIndex = 0;
          this.highlightedPlaceIndex = 0;
          this.currentUnitIndex = 0;
          this.currentReverseGeocodeLocationIsValid = true;
          this.intent = 'LATLNG';
        } else {
          this.places = [];
          this.currentPlaceIndex = 0;
          this.highlightedPlaceIndex = 0;
          this.currentUnitIndex = 0;
          this.currentReverseGeocodeLocationIsValid = false;
        }
      }),
      action('reverseGeocode error', err => {
        this.searching = false;
        this.lastSearchError = err;
        window._opbeat && window._opbeat('captureException', err);
      })
    );
  }

  @computed
  get currentPlace(): ?SearchAddressPlace {
    return (
      (this.places &&
        this.places.length > this.currentPlaceIndex &&
        this.places[this.currentPlaceIndex]) ||
      null
    );
  }

  @computed
  get currentUnit(): ?AddressUnit {
    return (
      (this.currentPlace &&
        this.currentPlace.units.length > this.currentUnitIndex &&
        this.currentPlace.units[this.currentUnitIndex]) ||
      null
    );
  }

  @computed
  get notFound(): boolean {
    return !!this.places && this.places.length === 0;
  }

  @computed
  get address(): string {
    if (this.currentUnit) {
      return this.currentUnit.address;
    } else if (this.currentPlace) {
      return this.currentPlace.address;
    } else {
      return '';
    }
  }

  @computed
  get addressId(): ?string {
    if (this.currentUnit) {
      return this.currentUnit.addressId;
    } else if (this.currentPlace) {
      return this.currentPlace.addressId;
    } else {
      return null;
    }
  }

  @computed
  get units(): Array<AddressUnit> {
    return this.currentPlace ? this.currentPlace.units : [];
  }
}
