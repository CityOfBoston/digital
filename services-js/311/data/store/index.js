// @flow

import { observable, computed, useStrict } from 'mobx';
import type { IPromiseBasedObservable } from 'mobx-utils';

import type { Service, ServiceSummary, SubmittedRequest } from '../types';
import Question from './Question';

// MobX will enforce that state changes only happen in action blocks.
useStrict(true);

export class ContactInfo {
  @observable required: boolean = false;
  @observable firstName: string = '';
  @observable lastName: string = '';
  @observable email: string = '';
  @observable phone: string = '';

  // can be false even when required is false to differentiate between submitting
  // and skipping
  @computed
  get requirementsMet(): boolean {
    return !!(this.firstName && this.lastName);
  }
}

export class LocationInfo {
  @observable required: boolean = false;
  @observable address: string = '';
  @observable.shallow location: ?{| lat: number, lng: number |} = null;

  // can be false even when required is false to differentiate between submitting
  // and skipping
  @computed
  get requirementsMet(): boolean {
    return this.address.length > 0;
  }
}

export class AppStore {
  @observable description: string = '';

  @observable contactInfo: ContactInfo = new ContactInfo();
  @observable locationInfo: LocationInfo = new LocationInfo();
  @observable questions: Question[] = [];

  @observable.shallow serviceSummaries: ServiceSummary[];
  serviceCache: Map<string, Service> = observable.shallowMap({});

  apiKeys: {[service: string]: string} = {};
  isPhone: boolean = false;

  @observable requestSubmission: ?IPromiseBasedObservable<SubmittedRequest> = null;

  @observable.ref _currentService: ?Service = null;
  @observable.ref currentServiceError: ?Object = null;

  @computed get currentService(): ?Service {
    return this._currentService;
  }

  set currentService(service: ?Service) {
    if (service === this._currentService) {
      return;
    }

    this._currentService = service;
    this.currentServiceError = null;
    this.questions = [];

    if (service) {
      try {
        this.contactInfo.required = service.contactRequired;
        this.locationInfo.required = service.locationRequired;
        this.questions = Question.buildQuestions(service.attributes);
      } catch (e) {
        this.currentServiceError = e;
      }
    }
  }

  @computed
  get questionRequirementsMet(): boolean {
    return this.questions.filter((q) => !q.required || !q.visible || q.requirementsMet).length === this.questions.length;
  }
}

let browserStore: AppStore;

/**
 * Returns a Redux store for this application. For the server, always returns a
 * fresh store with the given initialState. On the browser, will always return
 * the same store and initialState is ignored after the first call.
 */
export default function getStore(): AppStore {
  if (process.browser && browserStore) {
    return browserStore;
  }

  const store = new AppStore();

  if (process.browser) {
    browserStore = store;
  }

  return store;
}
