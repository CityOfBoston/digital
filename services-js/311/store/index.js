// @flow
/* global window */
/* eslint no-underscore-dangle: 0 */

import { createStore, compose, applyMiddleware } from 'redux';
import type { Store as ReduxStore } from 'redux';
import { middleware as reduxPackMiddleware } from 'redux-pack';
import thunk from 'redux-thunk';
import inject from 'redux-inject';

import type { RequestAdditions } from '../server/next-handlers';
import { reducer } from './modules';
import type { AppState, AppAction } from './modules';
import makeLoopbackGraphql from './deps/loopback-graphql';
import type { LoopbackGraphql } from './deps/loopback-graphql';

// Re-export these to keep things centralized in the “store” module.
export type State = AppState;
export type Action = AppAction;

export type Store = ReduxStore<State, Action>;

export type Deps = {|
  loopbackGraphql: LoopbackGraphql,
|}

// We implement our own Dispatch type to account for the actions and return
// values provided by our middleware. (The Redux built-in type is from Action
// (which must have a type property) to Action. We need to take Actions that
// don't have properties and return pretty much anything, since that’s what
// thunks can return.
export type Dispatch = (action: Action) => any;

// For the browser, we keep a singleton store that can be re-used across pages.
// A module global is the only mechanism for sharing between Next.js pages.
let browserStore = null;

// Factored out to allow for injecting test dependencies and state while
// maintaining the same middleware.
export function makeStore(deps: ?Deps = null, initialState: ?State = null): Store {
  // Support for Redux Devtools Extension: https://github.com/zalmoxisus/redux-devtools-extension
  const composeEnhancers = (process.browser && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;

  const enhancers = composeEnhancers(
    applyMiddleware(
      inject(deps),
      thunk,
      reduxPackMiddleware,
    ),
  );

  if (initialState) {
    return createStore(reducer, initialState, enhancers);
  } else {
    return createStore(reducer, enhancers);
  }
}

/**
 * Returns a Redux store for this application. For the server, always returns a
 * fresh store with the given initialState. On the browser, will always return
 * the same store and initialState is ignored after the first call.
 *
 * req is the Node request (expected to contain injected dependencies for server
 * data fetching) that should be provided when creating the store for the
 * server’s getInitialProps calls. Can be null when server rendering or on the
 * client.
 */
export default function getStore(req: ?RequestAdditions, initialState: ?State = null): Store {
  if (process.browser && browserStore) {
    return browserStore;
  }

  const deps = {
    loopbackGraphql: makeLoopbackGraphql(req),
  };

  const store = makeStore(deps, initialState);

  if (process.browser) {
    browserStore = store;
  }

  return store;
}
