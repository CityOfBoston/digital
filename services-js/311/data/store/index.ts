/* global window */

import { observable, configure as mobxConfigure, action } from 'mobx';
import Router from 'next/router';

import RequestSearch from './RequestSearch';
import AddressSearch from './AddressSearch';
import Ui from './Ui';
import BrowserLocation from './BrowserLocation';
import AllServices from './AllServices';
import {
  FetchGraphql,
  ScreenReaderSupport,
  GaSiteAnalytics,
  RouterListener,
} from '@cityofboston/next-client-common';
import { SiteAnalytics } from '@cityofboston/next-client-common/build/SiteAnalytics';

// MobX will enforce that state changes only happen in action blocks.
mobxConfigure({ enforceActions: true });

export type LanguagePreference = {
  code: string;
  region: string | null;
  quality: number;
};

export class AppStore {
  screenReaderSupport: ScreenReaderSupport = new ScreenReaderSupport();
  requestSearch: RequestSearch = new RequestSearch();
  ui: Ui = new Ui();
  browserLocation: BrowserLocation = new BrowserLocation();
  addressSearch: AddressSearch = new AddressSearch();
  allServices: AllServices = new AllServices();
  siteAnalytics: SiteAnalytics = new GaSiteAnalytics();

  languages: LanguagePreference[] = [];

  @observable
  liveAgentAvailable: boolean = (typeof window !== 'undefined' &&
    window.LIVE_AGENT_AVAILABLE) ||
  false;

  _liveAgentButtonId: string = '';

  get liveAgentButtonId(): string {
    return this._liveAgentButtonId;
  }

  set liveAgentButtonId(liveAgentButtonId: string) {
    this._liveAgentButtonId = liveAgentButtonId;

    if (typeof window !== 'undefined') {
      if (window.liveagent) {
        this.listenForLiveAgentEvents(liveAgentButtonId);
      } else {
        if (!window._laq) {
          window._laq = [];
        }

        window._laq.push(() =>
          this.listenForLiveAgentEvents(liveAgentButtonId)
        );
      }
    }
  }

  listenForLiveAgentEvents(liveAgentButtonId: string) {
    window.liveagent!.addButtonEventHandler(
      liveAgentButtonId,
      this.liveAgentEventHandler
    );
  }

  @action.bound
  liveAgentEventHandler(event: string) {
    switch (event) {
      case 'BUTTON_AVAILABLE':
        this.liveAgentAvailable = true;
        break;
      case 'BUTTON_UNAVAILABLE':
        this.liveAgentAvailable = false;
        break;
      default:
        break;
    }
  }
}

let browserStore: AppStore;

/**
 * Returns a Redux store for this application. For the server, always returns a
 * fresh store with the given initialState. On the browser, will always return
 * the same store and initialState is ignored after the first call.
 */
export default function getStore(
  fetchGraphql: FetchGraphql | null = null
): AppStore {
  if (process.browser && browserStore) {
    return browserStore;
  }

  const store = new AppStore();

  if (process.browser) {
    browserStore = store;
    store.screenReaderSupport.attach();
    store.ui.attach();
    store.browserLocation.attach(fetchGraphql);

    new RouterListener().attach({
      router: Router,
      siteAnalytics: store.siteAnalytics,
      screenReaderSupport: store.screenReaderSupport,
    });
  }

  return store;
}
