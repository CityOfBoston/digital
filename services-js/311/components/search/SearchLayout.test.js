// @flow

import React from 'react';
import renderer from 'react-test-renderer';

import { AppStore } from '../../data/store';
import type { Request } from '../../data/types';
import { makeServerContext } from '../../lib/test/make-context';

import SearchLayout from './SearchLayout';

export const MOCK_REQUEST: Request = {
  id: '17-000000001',
  service: {
    name: 'Cosmic Intervention',
  },
  description: 'I think that Thanos is here',
  status: 'open',
  address: 'City Hall Plaza, Boston, MA 02131',
  requestedAtString: 'March 7, 2017, 12:59 PM',
  updatedAtString: 'April 8, 2017, 12:59 PM',
};

describe('search form', () => {
  let data;

  beforeEach(async () => {
    const ctx = makeServerContext('/lookup', { q: 'Alpha Flight' });
    data = (await SearchLayout.getInitialProps(ctx)).data;
  });

  test('getInitialProps', () => {
    expect(data.view).toEqual('search');
    expect(data.query).toEqual('Alpha Flight');
  });

  test('rendering', () => {
    const component = renderer.create(
      <SearchLayout store={new AppStore()} data={data} />,
      {
        createNodeMock: () => document.createElement('div'),
      },
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
