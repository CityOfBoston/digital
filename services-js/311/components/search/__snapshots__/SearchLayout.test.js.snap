// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search form rendering 1`] = `
<div
  className="mn mn--full mn--nv-s css-1pujyrn"
  style={
    Object {
      "backgroundColor": "transparent",
    }
  }
>
  <nav
    className="nv-s nv-s--y nv-s--sticky"
  >
    <input
      aria-hidden={true}
      className="nv-s-tr"
      id="nv-s-tr"
      type="checkbox"
    />
    <ul
      className="nv-s-l"
    >
      <li
        className="nv-s-l-i"
      >
        <label
          className="nv-s-l-b"
          htmlFor="nv-s-tr"
          type="button"
        >
          Navigation
        </label>
      </li>
      <li
        className="nv-s-l-i"
      >
        <a
          className="nv-s-l-a "
          href="/"
          onClick={[Function]}
        >
          Report a problem
        </a>
      </li>
      <li
        className="nv-s-l-i"
      >
        <a
          className="nv-s-l-a nv-s-l-a--active"
          href="/search"
          onClick={[Function]}
        >
          Search cases
        </a>
      </li>
      <li
        className="nv-s-l-i"
      >
        <a
          className="nv-s-l-a "
          href="/services"
          onClick={[Function]}
        >
          All services
        </a>
      </li>
      <li
        className="nv-s-l-i"
      >
        <a
          className="nv-s-l-a "
          href="/faq"
          onClick={[Function]}
        >
          FAQ
        </a>
      </li>
    </ul>
  </nav>
  <div
    className={
      Object {
        "data-css-pmaldw": "",
      }
    }
  >
    <div
      className={
        Object {
          "data-css-wi5peg": "",
        }
      }
      style={
        Object {
          "opacity": 1,
        }
      }
    />
  </div>
  <div
    className={
      Object {
        "data-css-1g3oppu": "",
      }
    }
  >
    <div
      className="p-a300"
    >
      <form
        acceptCharset="UTF-8"
        action="/lookup"
        className="sf sf--y sf--md"
        method="get"
        onSubmit={[Function]}
      >
        <div
          className="sf-i"
        >
          <input
            className="sf-i-f"
            name="q"
            onChange={[Function]}
            placeholder="Search by case ID or keywordsâ€¦"
            type="text"
            value="Alpha Flight"
          />
          <button
            className="sf-i-b"
          >
            Search
          </button>
        </div>
      </form>
    </div>
    <div
      className={
        Object {
          "data-css-qtfk4q": "",
        }
      }
    >
      <div
        className={
          Object {
            "data-css-z2og4g": "",
          }
        }
      >
        <span
          className={
            Object {
              "data-css-1vjs1z7": "",
            }
          }
        >
          <svg
            className={
              Object {
                "data-css-mfy564": "",
              }
            }
            role="img"
          >
            <use
              height="100%"
              xlinkHref="/static/img/svg/loading-buildings.svg#east_boston"
            />
          </svg>
        </span>
      </div>
    </div>
  </div>
</div>
`;
